### ISSUE TITLE
The error exits when I export the compressed model to tflite after I used the uniform-tf learner.

### ISSUE THREAD

**Issue by yupanpanwangrui** [Tencent/PocketFlow#135]
After I compressed resnet20 using cifar10 dataset, I want to export it to pb and tflite.
I use the command:
python tools/conversion/export_quant_tflite_model.py 
The error is 
File "tools/conversion/export_quant_tflite_model.py", line 274, in <module>
    tf.app.run()
  File "/home/panpan/anaconda3/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "tools/conversion/export_quant_tflite_model.py", line 252, in main
    input_name, input_shape = get_input_name_n_shape(file_path_meta)
  File "tools/conversion/export_quant_tflite_model.py", line 63, in get_input_name_n_shape
    net_input = tf.get_collection(FLAGS.input_coll)[0]
IndexError: list index out of range

**Comment by yupanpanwangrui** [comment]
@jiaxiang-wu my issue is not resolved. In the learners/uniform_quantization_tf/learner.py, 278 line has 

tf.add_to_collection('images_final', images)
tf.add_to_collection('logits_final', logits)

so the tensors has added to the collections. I don't know why the compressed model can't to convert to pb or tflite. In your results, can it convert to pb or tflite using uniform-tf learner?

**Comment by jiaxiang-wu** [comment]
@yupanpanwangrui 
Can you provide the execution command for the model conversion script `tools/conversion/export_quant_tflite_model.py`?

**Comment by yupanpanwangrui** [comment]
I use
python tools/conversion/export_quant_tflite_model.py

**Comment by yupanpanwangrui** [comment]
I successed by using python tools/conversion/export_quant_tflite_model.py ----enbl_post_quant. Cheers!
why use python tools/conversion/export_quant_tflite_model.py failed?

**Comment by yupanpanwangrui** [comment]
I use resnet20 on cifar10 dataset. I don't set --model_dir, because I set it in the export_quant_tflite_model.py. "./models" directiory saves the compressed model. The files under the "./models" directory are:
checkpoint
model.ckpt.data-00000-of-00001
model.ckpt.index
model.ckpt.meta
They are renamed by me.
When  convert to tflite, exits

panpan@panpan-desktop:~/downloads/PocketFlow-master$ python tools/conversion/export_quant_tflite_model.py --enbl_post_quant
/home/panpan/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-12-08 16:26:03.635268: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-12-08 16:26:03.696129: W tensorflow/core/graph/graph_constructor.cc:1263] Importing a graph with a lower producer version 26 into an existing graph with producer version 27. Shape inference will have run different parts of the graph with different producer versions.
INFO:tensorflow:Restoring parameters from /home/panpan/downloads/PocketFlow-master/models/model.ckpt
INFO:tensorflow:input: net_input:0 / output: net_output:0
INFO:tensorflow:input's shape: (?, 32, 32, 3)
INFO:tensorflow:output's shape: (?, 10)
INFO:tensorflow:Restoring parameters from /home/panpan/downloads/PocketFlow-master/models/model.ckpt
INFO:tensorflow:Froze 174 variables.
INFO:tensorflow:Converted 174 variables to const ops.
INFO:tensorflow:/home/panpan/downloads/PocketFlow-master/models/model_original.pb generated
INFO:tensorflow:/home/panpan/downloads/PocketFlow-master/models/model_original.pb -> /home/panpan/downloads/PocketFlow-master/models/model_original.tflite
/home/panpan/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-12-08 16:26:07.301873: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
INFO:tensorflow:/home/panpan/downloads/PocketFlow-master/models/model_original.tflite generated
INFO:tensorflow:/home/panpan/downloads/PocketFlow-master/models/model_original.pb -> /home/panpan/downloads/PocketFlow-master/models/model_quantized.tflite
/home/panpan/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-12-08 16:26:14.406488: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
INFO:tensorflow:/home/panpan/downloads/PocketFlow-master/models/model_quantized.tflite generated
INFO:tensorflow:input: import/net_input:0 / output: import/net_output:0
INFO:tensorflow:outputs from the *.pb model: [[1.7683977e-01 1.4412816e-03 6.9048119e-01 1.0624725e-02 1.3871575e-02
  1.4870088e-03 9.7604260e-02 6.7596365e-04 5.8231270e-03 1.1511213e-03]]
INFO:tensorflow:input details: [{'name': 'net_input', 'index': 0, 'shape': array([ 1, 32, 32,  3], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
INFO:tensorflow:output details: [{'name': 'net_output', 'index': 1, 'shape': array([ 1, 10], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
INFO:tensorflow:outputs from the *.tflite model: [[1.7683969e-01 1.4412819e-03 6.9048125e-01 1.0624722e-02 1.3871565e-02
  1.4870083e-03 9.7604193e-02 6.7596370e-04 5.8231275e-03 1.1511213e-03]]
INFO:tensorflow:input details: [{'name': 'net_input', 'index': 0, 'shape': array([ 1, 32, 32,  3], dtype=int32), 'dtype': <class 'numpy.uint8'>, 'quantization': (0.007874015718698502, 128)}]
INFO:tensorflow:output details: [{'name': 'net_output', 'index': 1, 'shape': array([ 1, 10], dtype=int32), 'dtype': <class 'numpy.uint8'>, 'quantization': (0.00390625, 0)}]
INFO:tensorflow:outputs from the *.tflite model: [[ 0  0  0  2 62 62 62  6  0 62]]

The generated model_orginal.pb, model_original.tflite, model_quantized.tflite in the "./models" directiory.  The model_original.tflite is 1.1M, and model_quantized.tflite is 300.1KB. 

Is it Normal?

**Comment by jiaxiang-wu** [comment]
1. TensorFlow's quantization-aware training API cannot recognize all the operations that should be quantized in ResNet models (but for MobileNet, all operations are correctly recognized). Therefore, you need the `--enbl_post_quant` option to quantize the remaining operations for ResNet (but not needed for MobileNet).
2. You mentioned that files under the "./models" directory are renamed by you. What are their previous names? For exporting to TensorFlow Lite models, you need to provide checkpoint files of evaluation graph (instead of training graph), which are stored in the "./models_uqtf_eval" directory by default (the `uqtf_save_path_eval` option).

**Comment by yupanpanwangrui** [comment]
I don't have the "./models_uqtf_eval" directory. I don't know why. So I save the evaluation graph by myself when i set --exec_mode=eval.

**Comment by jiaxiang-wu** [comment]
How did you generate compressed models with `UniformQuantTFLearner`?

**Comment by yupanpanwangrui** [comment]
I add a line in the learners/uniform_quantization_tf/learner.py. In the function def evaluate(self): I add 

self.__save_model(is_train=False)
after self.__restore_model(is_train=False)

**Comment by yupanpanwangrui** [comment]
The execution command is 
./scripts/run_local.sh nets/resnet_at_cifar10_run.py --learner=uniform-tf exec_mode=eval
Then I don't have the "./models_uqtf_eval" directory.
I use the execution command to get tflite,

python tools/conversion/export_quant_tflite_model.py  
--model_dir ./models_uqtf_eval --enbl_post_quant
Success.

**Comment by jiaxiang-wu** [comment]
Why are you using `--exec_mode=eval` to enter the evaluation mode? Have you trained a quantized models with `UniformQuantTFLearner`?

**Comment by yupanpanwangrui** [comment]
Yes, I have trained a quantized models using the execution command 
./scripts/run_local.sh nets/resnet_at_cifar10_run.py --learner=uniform-tf --nb_epochs_rat 0.2
Then I  have the "./models_uqtf" directory.

**Comment by jiaxiang-wu** [comment]
When training is done, both "./models_uqtf" and "./models_uqtf_eval" directories should be generated.

**Comment by yupanpanwangrui** [comment]
I think so, but the "./models_uqtf_eval" directory doesn't generated.

**Comment by jiaxiang-wu** [comment]
Are you using the latest version of PocketFlow? Can you re-run the training process?

**Comment by yupanpanwangrui** [comment]
I re-run the training process using the latest version. Success. Thank for your reply.

I have another question. How to calculate the inference time of pb or tflite as export_chn_tflite_model.py do? Do you want to add inference time into export_quant_tflite_model.py lately?

**Comment by jiaxiang-wu** [comment]
Yes, we are adding benchmark tools for *.pb and *.tflite models and a PR will be created in the next one or two days.

**Comment by yupanpanwangrui** [comment]
Thanks! Looking forward to your update.
